### ISSUE TITLE
Mail queues support

### ISSUE THREAD

**Issue by multun** [shacker/django-todo#41]
### What ?

It might sound a little bit insane, but I would like to implement mail queues support.
The usecase is quite simple: making sure every mail of a given mailbox gets an answer.

Whenever a new mail is received, if it's already registered somewhere, comment the existing issue with the plaintext version of the received email. Otherwise, create a new issue.

### How ?

- A new model representing an email would be added. It would be referenced from comment
- The body attribute of comments would become a property, fetching the content of the email when present, falling back to the text field otherwise.
- A completly separate worker would register emails

### Why ?

I couldn't find a good and simple mail + bug tracker combo, and I figured it isn't _that_ much added complexity.

### What it's not

 - additionnal email notification madness / embed response
 - a massive code reworks

**Comment by shacker** [comment]
This is interesting, @multun , and is certainly in line with how some of the more advanced bug trackers work. If we do it, I want to make sure it is non-intrusive and fully optional. I can see two new settings being needed:

- TODO_EMAIL_INTEGRATION = True/False
- TODO_EMAIL_INTEGRATION_ADDRESS = 'foo@example.com' 

All features related to this proposal would be contingent on the value of `TODO_EMAIL_INTEGRATION`.

The latter would be the mailbox to watch for either augmenting an existing ticket or creating a new one. Although you did say "separate worker process" which implies that a separate worker would push content *into* django-todo. In that case we wouldn't need to "watch" a mailbox at all, but we would need some sort of listener to receive signals from the external worker - what do you propose for that? Would the worker system be up to the implementer? Would you provide an example implentation to be included in this repo? 

I'm not sure I understand why you would need to introduce a new model. I would think that an incoming email just gets copied in as a new comment on the thread.

**Comment by multun** [comment]
> I'm not sure I understand why you would need to introduce a new model. I would think that an incoming email just gets copied in as a new comment on the thread.

In order to attach an email to an existing thread, some additionnal metadata has to be kept (the message id)
Furthermore, an email doesn't have the same attributes as a regular comment : no user created it, but it comes from a source address. The additionnal model would be there to avoid messing the comment model up.

> but we would need some sort of listener to receive signals from the external worker

Why ? the worker could directly push its things to the database.

> Would the worker system be up to the implementer

A worker is in charge of pushing emails from a mailbox into a task list.
It could be started that way :

`python -m django_todo.mail_worker -c /etc/djando_todo_workers/some_worker.yml`

`some_worker.yml`:
```yml
task_list: my_task_list
mail_source:
    type: imap
    user: someuser
    password: somepassword
    host: imap.example.com
    port: 4242
```

**Comment by multun** [comment]
@shacker PR #43 features a prototype implementation, what do you think?
I think you were right about just adding attributes to the comment, it currently isn't that great looking

**Comment by shacker** [comment]
> an email doesn't have the same attributes as a regular comment : no user created it, but it comes from a source address. 

I disagree with "no user created it." If the incoming email matches the email of a registered user, then we can assume that user created it, and it can be posted as a regular comment by that user. We would have to decide what to do with incoming emails that don't match the email of a registered user - I would assume we discard them, presumably with a "Sorry..." email response. 

Thanks for the PR, will check it out tomorrow.

**Comment by multun** [comment]
> We would have to decide what to do with incoming emails that don't match the email of a registered user

Discarding these would mostly defeat the purpose of email tracking, which in my case, would be tracking the progress of tasks submitted by email. With the current implementation, everyone is able to create a task and comment it, without even having to access a web interface. Even better, the end user has no idea the task even exists, so that regular task comments can be used to track the progress of the task without bothering the end user.

> If the incoming email matches the email of a registered user, then we can assume that user created it, and it can be posted as a regular comment by that user

That would be a cool feature. Do the user that created a comment have some additional privileges ? The email "From" header is user controlled. It isn't always authenticated.

**Comment by shacker** [comment]
OK I see where you're going with this now. Up to now I've been thinking about this feature as supporting an alternate path for registered users -- to get support via email rather than through the web UI (since that's a workflow I'm used to at work). The current `external_add` view is in fact an authenticated view:

```
@login_required
def external_add(request) -> HttpResponse:
    """Allow authenticated users who don't have access to the rest of the ticket system to file a ticket
    in the list specified in settings (e.g. django-todo can be used a ticket filing system for a school, where students can file tickets without access to the rest of the todo system).

    Publicly filed tickets are unassigned unless settings.DEFAULT_ASSIGNEE exists.
    """
```

So if I understand correctly, you also want to change this aspect of django-todo, making `external_add` truly open to the public.

I am not overly concerned about people forging From headers (could happen but not likely) but I would be very concerned about managing spam if accepting input from unknown users, which is why this was created as an authenticated view to begin with. What would you suggest as an anti-spam strategy 

OK, think of these scenarios and combinations and how django-todo would handle them:

**Web UI**
1) Logged-in user filing new ticket via web
2) Logged-in user adding comment to existing ticket via web
3)  Non-auth user submitting new ticket via web
4) Non-auth user submitting comment via web

**Email**
5)  Known user (email matches a registered user) submitting new ticket via email
6) Known user (email matches a registered user) adding a response/comment to existing ticket via email
7) Unknown user submitting new ticket via email
8) Unknown user responding to comment thread via email

Obviously 1 and 2 are already covered, but I think we need to lay out the internal workflow and logic for 3-8 clearly in advance, so that the system doesn't become a spam magnet. 

Think about this as well: Let's say an inbound email references a ticket ID in the subject and we use that to determine which ticket to attach the comment to. What would stop a spammer from creating emails that reference existing ticket IDs that are not theirs, thus submitting spam as comments to other people's threads? Since they're not known users, seems like we need some other layer of protection. Ideas?

**Comment by multun** [comment]
> you also want to change this aspect of django-todo, making external_add truly open to the public.

Well not really, but if you want to why not?

> What would you suggest as an anti-spam strategy

I don't think anything can be done at all when it comes to emails ? filtering is done upstream already. The emails are pulled from the INBOX folder by default, messages classified as junk don't get there at all.

> What would stop a spammer from creating emails that reference existing ticket IDs that are not theirs, thus submitting spam as comments to other people's threads?

The fact they have to know the message id of a message from an existing thread, which doesn't sound realistic when it comes to spam. We don't use ticket IDs, but message IDs. These are super long strings identifying an email. I don't think you can guess these.

> 5) Known user (email matches a registered user) submitting new ticket via email
> 6) Known user (email matches a registered user) adding a response/comment to existing ticket via email

How about keeping this simple ? I still don't understand why these should be different from emails from unknown senders. I don't even see what's the additional benefit of having a user bound to an email comment.

What if somebody spoofs the email address of somebody important, and files a ticket which asks for something critical?

> Unknown user submitting new ticket via email
> Unknown user responding to comment thread via email

The two are the primary reasons why this issue exists: Think of this feature as a shared mailbox where every thread is a new issue internal staff can comment on.

**Comment by shacker** [comment]
I think we're using "spam" in different context. I'm not referring to typical email spam finding its way into the system, but more like weblog spam, where a sender's goal is to get their ad or URL into your system. If a person knows that all they have to do is send an email advertising Joe's Garage to foo@example.com and it will appear in the ticket list for all users of the system, that's going to be exploited. For comments on existing threads, we can match on message IDs. But for new issues, there is no message ID to match to and it goes straight in. To me this feels like an open barn door. 

Thinking back on my experience with this sort of thing in the past, the communication does not typically begin with me sending a plain email to support@example.com. Instead, I'm required to log into their site (thus becoming a known user with a known email address), then I use a form to Contact Support. At that point, a ticket exists attached to me, and we can conduct the rest of the communication via email - as a user I never have to visit the site again, though the entire correspondence is tracked in the ticket thread.

I would be very open to an architecture like that, but I am extremely reluctant about the idea of creating a ticket from a plain, unverified email. 

WDYT?

**Comment by multun** [comment]
> To me this feels like an open barn door.

Well creating tickets for new emails is the point of this feature. It isn't the regular mail tracking other bugtrackers might do: the mail tracking feature isn't there to track the progress of the conversation, it's the other way around.

Once an organization reaches a critical mass of about 25 people, it starts to get very hard to keep track of inbound emails without forgetting some, or answering multiple times.

It is an open door the same way a mailbox is.

> Thinking back on my experience with this sort of thing in the past, the communication does not typically begin with me sending a plain email to support@example.com

That's exactly the reason why this PR exists: I couldn't find any open source project tracking emails.

> Instead, I'm required to log into their site (thus becoming a known user with a known email address), then I use a form to Contact Support.

Well, tracking emails the way I would like to also have the benefit of enabling tracking of _outbound_ emails. If some people in my organization sends an email, it can CC the tracked email address, and people can comment about the progress of the conversation without bothering the recipient of the email.

> I would be very open to an architecture like that, but I am extremely reluctant about the idea of creating a ticket from a plain, unverified email.

I just think of it as a way to turn django-todo into a shared mailbox with threading, comments and more

**Comment by shacker** [comment]
OK, I finally understand your vision. I was concerned that anyone visiting the bug tracker would see any inbound spams, but of course the "tickets" queue (or whatever) would only be visible to administrators of that queue. Therefore you're right that it's very similar to using a shared mailbox where all admins of that box see the same spam, but no one else does. 

How about this: Introduce another setting `TODO_ALLOW_NEW_TICKETS_VIA_EMAIL`. We would still accept new comments regardless of that setting because we have the message_id working as gatekeeper for those, but we would only create new tickets via email if this setting is True. This would be part of the documentation you're about to write :) Sound OK? 

We should also think about how we're going to demo this on the demo site (or maybe we don't have to). If we do want to, I'll need to set up a mailbox, start your worker on the server, and get things configured to let people experiment with it. Or maybe it's enough to document it and not make it part of the live demo. Thoughts?

**Comment by multun** [comment]
> but we would only create new tickets via email if this setting is True

How is that different from not starting a worker? Unless an initial email is sent when a ticket is created, there would be no way to even comment.

> This would be part of the documentation you're about to write :) 

o/

> I'll need to set up a mailbox, start your worker on the server, and get things configured to let people experiment with it. Or maybe it's enough to document it and not make it part of the live demo. Thoughts?

I have no idea. You're the one that would have to do it in the end, so it's all up to you!

**Comment by shacker** [comment]
> How is that different from not starting a worker?

No I'm talking about *with* the worker running. Implementers running the worker would, I think, be most likely to be willing to accept comments via email on existing tickets, but would not be willing to have new tickets *created* by simply sending an email. I don't think it's a safe assumption that just because the worker is running, implementers are willing to open it up that wide. Because of its potential for accepting garbage tickets into the system via email, it should be a conscious choice on the part of the implementer.

**Comment by multun** [comment]
Then how would you like emails to be attached to existing tickets? Currently the algorithm relies on the message id of existing email comments. If there's no email comment in the first place, you can't attach new comments.

> Because of its potential for accepting garbage tickets into the system via email, it should be a conscious choice on the part of the implementer.

It would. The current version of the mail tracker is configured as a producer / consumer pattern. Currently, there's only one consumer, which is a mail tracker. I would like to make it as clear as possible in the documentation.

**Comment by shacker** [comment]
@multun Sorry to leave you hanging for so long! Returned from holiday to some work pressures (which are ongoing). 

> I should be easy to add another consumer which implements another behavior / add an optionnal parameter to tweak how this one behaves, either later on or now.

OK, that makes sense to me. Thanks for thinking this through so carefully. Go ahead and implement it as you've envisioned, then we can modify / add another consumer later, as needed. Can you get full documentation into the next commit, and we'll see how close to ready we are. Thanks!

**Comment by shacker** [comment]
@multun Just checking in - did you intend to continue this work, or should I close this issue? I think our philosophies on it differed a bit, so I'm not concerned if you want to continue the work on your own fork. Just let me know please.

**Comment by multun** [comment]
@shacker I actually intend to continue this work (I even believe it's now ready). I'll be updating my pull request in the coming days.

> I think our philosophies on it differed a bit, so I'm not concerned if you want to continue the work on your own fork.

I don't think the internet needs one more fork ;-)

**Comment by shacker** [comment]
2.3 is released to pypi, with mail tracker and CSV import support. Just wanted to thank you again, @multun, for this excellent contribution. Cheers!
